lib/cloudex/cloudinary_api.ex

```diff
@@ -92,33 +92,53 @@ defmodule Cloudex.CloudinaryApi do
   @spec upload_file(String.t(), map) ::
           {:ok, %Cloudex.UploadedImage{}} | {:ok, %Cloudex.UploadedVideo{}} | {:error, any}
   defp upload_file(file_path, opts) do
+    poison_opts = get_httpoison(opts)
+
     options =
       opts
       |> extract_cloudinary_opts
+      |> extract_poison_opts
       |> prepare_opts
       |> sign
       |> unify
       |> Map.to_list()
 
     body = {:multipart, [{:file, file_path} | options]}
 
-    post(body, file_path, opts)
+    post(body, file_path, opts, poison_opts)
   end
 
   @spec extract_cloudinary_opts(map) :: map
   defp extract_cloudinary_opts(opts) do
     Map.delete(opts, :resource_type)
   end
 
+  @spec get_httpoison(map) :: map
+  defp get_httpoison(opts) do
+    if Map.has_key?(opts, :httpoison) do
+      Map.get(opts, :httpoison)
+    else
+      []
+    end
+  end
+
+  @spec extract_poison_opts(map) :: map
+  defp extract_poison_opts(opts) do
+    Map.delete(opts, :httpoison)
+  end
+
   @spec upload_url(String.t(), map) ::
           {:ok, %Cloudex.UploadedImage{}} | {:ok, %Cloudex.UploadedVideo{}} | {:error, any}
   defp upload_url(url, opts) do
+    poison_opts = get_httpoison(opts)
+
     opts
+    |> extract_poison_opts
     |> Map.merge(%{file: url})
     |> prepare_opts
     |> sign
     |> URI.encode_query()
-    |> post(url, opts)
+    |> post(url, opts, poison_opts)
```

Comment:
I think even with calling `Map.delete`, the `opts` map shouldn't change, so it should still have the `:httpoison` key. maybe instead of passing in poison_opts here, we could keep it `post(url, opts)` and then in `common_post` we could do `poison_opts = get_httpoison(opts)` and then pass those into the actual httparty request

Reply:
This should be sufficient:

```elixir
{request_opts, opts} = Map.pop(opts, :request_options, [])
```